libcontainer/configs/cgroup.go

```diff
@@ -57,6 +57,11 @@ type Cgroup struct {
 	// MEM to use
 	CpusetMems string `json:"cpuset_mems"`
 
+	// Process limit; set to `0' to disable limit.
+	// While technically `0' is a valid PID limit, it does not make sense in the
+	// context of a container -- it is identical to a limit of `1'.
```

Comment:
@yangdongsheng 

> ```
> pids.max = N (N > 0) means there are at most N pids in this group, only check in copy_process().
> pids.max = 0 means No more copy_process() is allowed here.
> ```

That distinction does not exist within the PIDs cgroup controller. It treats `0` like any other limit, and it just so happens that `0` is **precisely** (note the use of the word **precisely**) identical to a limit of `1`, since you can't impose a limit on `0` processes... Yes, it would make a difference **IF** we blocked attach, but we don't and so there isn't a distinction.

> BTW, @cyphar why did you only check pids limit in copy_process(), ignoring to check attaching new pids and setting pids.max? This design would cause pids.current > pids.max which is really confusing.

The original design had limits set for attaching as well. However, there are two main reasons that attaching is not forbidden. The first is that being able to attach without following the limit doesn't compromise the limiting capability, as you must first have a process be spawned *somewhere* in order to attach it, meaning that you must be able to fork in some cgroup anyway and therefore you must've not broken some limit. Secondly is that making process limiting block attaches means that you can't attach a process to figure out what's going on, which makes management basically impossible. Also, Tejun has strong views about cgroup core blocking attach for the above reason. All of these reasons *are* outlined in the documentation (or in various mailing list posts), and the current design works perfectly fine IMO.